Update Django to 3.2

maurizi opened the pull request:
## Overview

Updates Django to 3.2 & Python to 3.7. Also updates a number of Django-related libraries, as well as a few libraries that required updating to meet the new Python version.

I did _not_ update `dedupe`, which is why we're on Python 3.7 (the latest version supported by the version of `dedupe` we're using). We'll have to defer further Python updates until we update `dedupe`.

Connects #1908 

## Notes

My process was as follows:
 - I read through the Django release notes and checked for any backwards incompatible changes
 - I updated the underlying Dockerfile to use the latest versions of Python & Django (I later backdated the Python version)
 - I went through the CHANGELOG for each requirement in our `requirements.txt` to see if they needed an update based on either the new Django version or the new Python version (this mostly matched Tai's investigation but I spotted a few more that needed updating)
 - `scripts/update`
 - I added `-W all` to the call to `python manage.py` in `scripts/manage`, and then looked for warnings / errors in:
   - `scripts/manage check`
   - `scripts/manage test`
 - Similarly I added `entrypoint: "python -Wa /usr/local/bin/gunicorn"` under the `django` container in the `docker--compose.yml` file and ran `scripts/server`
 - You can ignore warnings from specific modules, so after deciding it was OK to ignore and/or infeasible to fix warnings from these, I got to 0 warnings w/ these arguments: `-Wa -W ignore:::thefuzz.fuzz-W ignore:::fastcluster -W ignore:::dedupe.blocking -W ignore:::dedupe.core -W ignore:::rest_auth.urls -W ignore:::rest_auth.registration.urls`
   - `thefuzz` & `fastcluster` are both related to `dedupe`, which I mentioned above.
   - the `rest_auth` warnings aren't fatal (yet) so I deferred them since we'll likely have to migrate away from the library entirely

In the end my code changes to check for warnings ended up like this:
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 2402c0c..1dfa757 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -37,6 +37,7 @@ services:
     depends_on:
       database:
         condition: service_healthy
+    entrypoint: "python -Wa -W ignore:::fastcluster -W ignore:::dedupe.blocking -W ignore:::rest_auth.urls -W ignore:::rest_auth.registration.urls /usr/local/bin/gunicorn"
     command:
       - "-b :8081"
       - "--reload"
diff --git a/scripts/manage b/scripts/manage
index f0ccd7d..855bb26 100755
--- a/scripts/manage
+++ b/scripts/manage
@@ -20,6 +20,6 @@ if [ "${BASH_SOURCE[0]}" = "${0}" ]; then
         docker-compose \
             run --rm --entrypoint python \
             django \
-            manage.py "$@"
+            -Wa -W ignore:::fastcluster -W ignore:::dedupe.blocking -W ignore:::dedupe.core -W ignore:::rest_auth.urls -W ignore:::rest_auth.registration.urls -W ignore:::thefuzz.fuzz manage.py "$@"
     fi
 fi
```

## Testing Instructions

* `scripts/update`
* Run the application and look for any bugs
* I don't know if we have code coverage set up on this project, but if so it might be good to look to that to see what is not covered by the test suite and focus on that for manual testing

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] This PR is targeted at the correct branch (`develop` vs. `ogr/develop`)
- ~[ ] If this PR applies to both OAR and OGR a companion PR has been created~

maurizi commented on src/django/api/aws_batch.py:
```diff
@@ -1,9 +1,9 @@
 import boto3
 import json
```
The changes in this commit are the most far-reaching in the PR - I'd appreciate a careful double-check that I'm doing the right thing here. I think this is mostly a 1-1 replacement of the use of naive UTC datetimes with identical datetimes that have their `tzinfo` set to UTC, but I always get wary when dealing w/ dates & timezones.

jwalgran commented:
From a fresh checkout `./scripts/update` is failing. I should be able to work around this to continue my review but wanted to see if you could duplicate it

```
~/Projects/open-supply-hub (update/mvm/django)
> ./scripts/update
database uses an image, skipping
app uses an image, skipping
Building django
[+] Building 4.4s (11/11) FINISHED
 => [internal] load build definition from Dockerfile                                                                                    0.0s
 => => transferring dockerfile: 969B                                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => => transferring context: 2B                                                                                                         0.0s
 => [internal] load metadata for quay.io/azavea/django:3.2-python3.7-slim                                                               0.6s
 => [1/7] FROM quay.io/azavea/django:3.2-python3.7-slim@sha256:aae10f8b9fd9fbd4d87d6606e26e9bb75c2e43dae7098a9d4fcf19fe8743cafc         0.0s
 => [internal] load build context                                                                                                       0.1s
 => => transferring context: 772.12kB                                                                                                   0.0s
 => CACHED [2/7] RUN mkdir -p /usr/local/src                                                                                            0.0s
 => CACHED [3/7] WORKDIR /usr/local/src                                                                                                 0.0s
 => CACHED [4/7] COPY requirements.txt /usr/local/src/                                                                                  0.0s
 => CACHED [5/7] RUN set -ex     && buildDeps="     build-essential     "     && deps="     postgresql-client-13     "     && apt-get   0.0s
 => [6/7] COPY . /usr/local/src                                                                                                         0.2s
 => ERROR [7/7] RUN GOOGLE_SERVER_SIDE_API_KEY=""     OAR_CLIENT_KEY=""     MAILCHIMP_API_KEY=""     MAILCHIMP_LIST_ID=""     EXTERNAL  3.4s
------
 > [7/7] RUN GOOGLE_SERVER_SIDE_API_KEY=""     OAR_CLIENT_KEY=""     MAILCHIMP_API_KEY=""     MAILCHIMP_LIST_ID=""     EXTERNAL_DOMAIN=""     python manage.py collectstatic --no-input:
#11 3.247 /usr/local/lib/python3.7/site-packages/thefuzz/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
#11 3.247   warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')
#11 3.248 Traceback (most recent call last):
#11 3.248   File "manage.py", line 21, in <module>
#11 3.248     execute_from_command_line(sys.argv)
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
#11 3.248     utility.execute()
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 413, in execute
#11 3.248     self.fetch_command(subcommand).run_from_argv(self.argv)
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/core/management/base.py", line 354, in run_from_argv
#11 3.248     self.execute(*args, **cmd_options)
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/core/management/base.py", line 398, in execute
#11 3.248     output = self.handle(*args, **options)
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 187, in handle
#11 3.248     collected = self.collect()
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 105, in collect
#11 3.248     for path, storage in finder.list(self.ignore_patterns):
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/contrib/staticfiles/finders.py", line 130, in list
#11 3.248     for path in utils.get_files(storage, ignore_patterns):
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/contrib/staticfiles/utils.py", line 23, in get_files
#11 3.248     directories, files = storage.listdir(location)
#11 3.248   File "/usr/local/lib/python3.7/site-packages/django/core/files/storage.py", line 323, in listdir
#11 3.248     for entry in os.scandir(path):
#11 3.248 FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/src/static/static'
------
executor failed running [/bin/sh -c GOOGLE_SERVER_SIDE_API_KEY=""     OAR_CLIENT_KEY=""     MAILCHIMP_API_KEY=""     MAILCHIMP_LIST_ID=""     EXTERNAL_DOMAIN=""     python manage.py collectstatic --no-input]: exit code: 1
ERROR: Service 'django' failed to build : Build faile

jwalgran reviewed (approved):
Excellent work. I only found one broken item and suggested a simple one-line fix. I have a few inline questions as well. My raw test notes are below.

---

- /admin/reports/ is 404
	- Fix suggested
- datetime to timezone
	- We have had `USE_TZ=True`  set since we started, so switching to `timezone` is in line with our intention since the start
	- Our `TIME_ZONE` has always been set to `UTC`. Django has been using this to convert naïve datetimes ( https://docs.djangoproject.com/en/3.2/topics/i18n/timezones/#interpretation-of-naive-datetime-objects) Explicitly switching to `timezone.now()` is consistant
	- aws_batch.py
		- Dates are used to create string names and compute runtime by subtracting start time from end time. No problem.
	- limits.py
		- Made datetime aware. Looks good and covered by a test.
	- matching.py
		- Get start and finish times of jobs that are returned in JSON. Consistant and ok.
	- oar_id.py
		- We just use a datetime to get the day of the year. No problems.
	- processing.py
		- timestamps saves in json processing results as strings. No problems.
	- test.py
		- Made some previously naïve datetimes aware. No problems.
		- String timestamps in json results. No problems.
		- Sorting test ok because all datetimes are consistent. No problems
	- views.py
		- string timestamps in json processing results. ok.
		- status change dates. No problems.
- 0088_migrate_claim_fields.py
	- Change looks correct but will have already been run in all of our databases  so no problems


- Delete works

CN2022194HTZHWW

```
openapparelregistry=# select history_change_reason from api_historicalfacilitymatch where history_change_reason is not null;
  history_change_reason
-------------------------
 Deleted CN2022194HTZHWW
 (1 row)
```

- Merge works

US2022194VMHDS2
US2022194PMBKPT

```
openapparelregistry=# select history_change_reason from api_historicalfacility  where history_change_reason is not null;
    history_change_reason
-----------------------------
 Merged with US2022194VMHDS2
(1 row)

openapparelregistry=# select history_change_reason from api_historicalfacilitymatch where history_change_reason is not null;
            history_change_reason
---------------------------------------------
 Deleted CN2022194HTZHWW
 Merged US2022194PMBKPT into US2022194VMHDS2
(2 rows)
```

- Promote works

```
openapparelregistry=# select history_change_reason from api_historicalfacility  where history_change_reason is not null;
                history_change_reason
------------------------------------------------------
 Merged with US2022194VMHDS2
 Promoted item 336 in list 8 over item 630 in list 15
(2 rows)
```

- Split works
- Closed facility reporting worked
- Link to new  works

MX202219493RV4W
MX2022194YDBW7J

```
openapparelregistry=# select new_oar_id from api_facility where id = 'MX202219493RV4W';
   new_oar_id
-----------------
 MX2022194YDBW7J
(1 row)
```

- User registration works
- Profile update works
- password change works
- password reset works
- API token generation works
- Api block creation and limit enforcement works
- Submit facility by API works
- Claim workflow works
- Embedded map works

maurizi commented on src/django/Dockerfile:
```diff
@@ -9,7 +9,7 @@ RUN set -ex \
     " \
     && deps=" \
-    postgresql-client-12 \
+    postgresql-client-13 \
```
Updating to a new version of the Docker image updated the underlying OS, and thus the packages available via `apt-get` - we can no longer install `postgresql-client-12`.

maurizi commented on src/django/oar/settings.py:
```diff
@@ -303,6 +303,7 @@
 
 STATIC_URL = '/static/'
 STATIC_ROOT = os.path.join(BASE_DIR, 'static')
+STATICFILES_DIRS = ((os.path.join(STATIC_ROOT, "static")),)
```
I had just copied this from the MMW pull request adding `django-spa` w/o really understanding it's purpose, but removing it made the Django server stop serving the compiled Webpack bundle from `localhost:8081` - which is I believe how we serve static assets for staging/production?

maurizi commented:
@jwalgran this has been rebased to incorporate the `dedupe` update (which I then updated to the latest patch version: https://github.com/open-apparel-registry/open-apparel-registry/pull/1965/commits/7f68a81f76aa2c7ab45f1d6f9812d59ee8960ecd), I updated Python to 3.10, fixed the one deprecation warning that uncovered (https://github.com/open-apparel-registry/open-apparel-registry/pull/1965/commits/2752fd3a7903080020067207dd32e02c61bb35e7), and updated the linters to support Python 3.10 (https://github.com/open-apparel-registry/open-apparel-registry/pull/1965/commits/c5f730efe4a10107eda15f69138f25d983278ffb)

jwalgran reviewed (approved):
I verified that this runs in development and responded to the open comments. I think this is good. Thanks.
